lib/parser.ex

```diff
@@ -0,0 +1,205 @@
+defmodule Membrane.Ogg.Parser do
+  import Bitwise
+
+  @type packet :: %{payload: binary, track_id: integer, bos?: boolean, eos?: boolean}
+
+  @spec parse(binary, continued_packets :: %{}) ::
+          {parsed :: list(packet), unparsed :: binary, continued_packets :: %{}}
+  def parse(bytes, continued_packets) do
+    do_parse([], bytes, continued_packets)
+  end
+
+  @spec do_parse(acc :: list(packet), bytes :: binary, continued_packets :: %{}) ::
+          {parsed :: list(packet), bytes :: binary, continued_packets :: %{}}
+  defp do_parse(acc, bytes, continued_packets) do
+    case maybe_parse_page(bytes, continued_packets) do
+      {:need_more_bytes} ->
+        {acc, bytes, continued_packets}
+
+      {:ok, unparsed, packets, continued_packets} ->
+        do_parse(acc ++ packets, unparsed, continued_packets)
+    end
+  end
+
+  @spec maybe_parse_page(page :: binary, continued_packets :: %{}) ::
+          {:need_more_bytes}
+          | {:ok, bytes :: binary, packets :: list(packet), continued_packets :: %{}}
+  defp maybe_parse_page(initial_bytes, continued_packets) do
+    with {:ok, bytes, header_type, bitstream_serial_number, number_of_page_segments} <-
+           parse_header(initial_bytes),
+         {:ok, bytes, segment_table} <- parse_segment_table(bytes, number_of_page_segments),
+         {:ok} <- verify_length_and_crc(initial_bytes, segment_table) do
+      {bytes, packets, page_continued_packet} = parse_segments(bytes, segment_table)
+
+      {packets, continued_packets, page_continued_packet} =
+        prepend_continued_packet(
+          continued_packets,
+          bitstream_serial_number,
+          packets,
+          page_continued_packet
+        )
+
+      packets =
+        Enum.map(packets, fn packet ->
+          %{
+            payload: packet,
+            track_id: bitstream_serial_number,
+            bos?: (header_type &&& 0x2) > 0,
+            eos?: (header_type &&& 0x4) > 0
+          }
+        end)
+
+      continued_packets =
+        if page_continued_packet == nil do
+          continued_packets
+        else
+          Map.put(continued_packets, bitstream_serial_number, page_continued_packet)
+        end
+
+      {:ok, bytes, packets, continued_packets}
+    end
+  end
+
+  defp verify_length_and_crc(bytes, segment_table) do
+    segments_count = Enum.count(segment_table)
+    content_length = Enum.sum(segment_table)
+
+    if 22 + 4 + 1 + segments_count + content_length > byte_size(bytes) do
+      {:need_more_bytes}
+    else
+      <<before_crc::binary-size(22), crc::little-unsigned-size(32),
+        after_crc::binary-size(1 + segments_count + content_length), _rest::binary>> = bytes
+
+      crc_payload = before_crc <> <<0::size(32)>> <> after_crc
+
+      calculated_crc =
+        CRC.crc(
```

Comment:
There are some functions for calculating crc in the erlang standard library. Maybe we could rely on them? Take a look at:

https://www.erlang.org/doc/man/erlang.html#crc32-1

You call erlang code from the elixir as 

```elixir
 :module.function(...)
```

Reply:
I don't think so, ogg uses a weird variant of crc32:
```
iex(4)> :erlang.crc32(crc_payload)
1110206997
iex(5)> CRC.crc(
...(5)>           %{
...(5)>             extend: :crc_32,
...(5)>             init: 0x0,
...(5)>             poly: 0x04C11DB7,
...(5)>             xorout: 0x0,
...(5)>             refin: false,
...(5)>             refout: false
...(5)>           },
...(5)>           crc_payload
...(5)>         )
1675672756
```

Reply:
Eh :(